lib/compress/zstd_compress.c

```diff
@@ -2856,6 +2887,49 @@ void ZSTD_resetSeqStore(seqStore_t* ssPtr)
     ssPtr->longLengthType = ZSTD_llt_none;
 }
 
+/* ZSTD_postProcessExternalMatchFinderResult() :
+ * Validates and post-processes sequences obtained through the external matchfinder API.
+ * Returns the number of sequences after post-processing, or an error code. */
+static size_t ZSTD_postProcessExternalMatchFinderResult(
+    ZSTD_Sequence* outSeqs, size_t nbExternalSeqs, size_t outSeqsCapacity, int emptySrc
+) {
+    RETURN_ERROR_IF(
+        nbExternalSeqs > outSeqsCapacity,
+        externalMatchFinder_failed,
+        "External matchfinder returned error code %lu",
+        (unsigned long)nbExternalSeqs
+    );
+
+    RETURN_ERROR_IF(
+        nbExternalSeqs == 0 && !emptySrc,
+        externalMatchFinder_failed,
+        "External matchfinder produced zero sequences for a non-empty src buffer!"
+    );
+
```

Comment:
This method seems about validating that the output of the external matchfinder makes sense.

Another potential test that seems to make sense here
is to check that `sum(matchlen) + sum(litlen) == srcSize;`.
This is a heavy test, so that's only something we would want in debug mode, aka `assert()`.

Reply:
How can I gate the test so it only runs in debug mode? Is there an ifdef?

Reply:
Looks like this is the idiom?

```
#if defined(DEBUGLEVEL) && (DEBUGLEVEL>=2)
// code
#endif
```

Reply:
Yes, that method works.

Another method is to rely on DCE (Dead Code Elimination),
and you could have schematically something like `assert(test());`

For this case, you could write almost exactly `assert( sum(matchlen) + sum(litlen) == srcSize )`
or even `assert( sum(seqLen) == srcSize )`
provided you could provide the relevant `sum()`.

Reply:
I don't want to run the sum function unless we're in debug mode... and we don't have lambdas in C :) So I expect the `#ifdef` method is easier here.

Reply:
It turns out that Sen's code provides for sequence validation (both offsets and matchLens + litLens):
* https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_compress.c#L6004
* https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_compress.c#L6021

So we probably don't want to duplicate that code in `ZSTD_postProcessExternalMatchFinderResult`. I'll just make sure that debug builds have the validateSequences cParam enabled.

Reply:
Note: given that Sen's validation checks that `sum(matchLens) + sum(litLens) == srcSize` and only reads up to the block delimiter, we don't need any special check for `matchLen == 0` in postProcessSequences.

If the sequences returned by the external matchfinder have an "early" or "accidental" block delimiter which comes before the real end of the sequences, they will fail the summation test (since Sen's code will stop reading early).